aws/resource_aws_storagegateway_nfs_file_share.go

```diff
@@ -0,0 +1,392 @@
+						"directory_mode": {
+							Type:     schema.TypeString,
+							Optional: true,
+							Default:  "0777",
+						},
+						"file_mode": {
+							Type:     schema.TypeString,
+							Optional: true,
+							Default:  "0666",
+						},
+						"group_id": {
+							Type:         schema.TypeInt,
+							Optional:     true,
+							Default:      65534,
+							ValidateFunc: validation.IntBetween(0, 4294967294),
+						},
+						"owner_id": {
+							Type:         schema.TypeInt,
+							Optional:     true,
+							Default:      65534,
+							ValidateFunc: validation.IntBetween(0, 4294967294),
+						},
+					},
+				},
+			},
+			"object_acl": {
+				Type:     schema.TypeString,
+				Optional: true,
+				Default:  storagegateway.ObjectACLPrivate,
+				ValidateFunc: validation.StringInSlice([]string{
+					storagegateway.ObjectACLAuthenticatedRead,
+					storagegateway.ObjectACLAwsExecRead,
+					storagegateway.ObjectACLBucketOwnerFullControl,
+					storagegateway.ObjectACLBucketOwnerRead,
+					storagegateway.ObjectACLPrivate,
+					storagegateway.ObjectACLPublicRead,
+					storagegateway.ObjectACLPublicReadWrite,
+				}, false),
+			},
+			"read_only": {
+				Type:     schema.TypeBool,
+				Optional: true,
+				Default:  false,
+			},
+			"requester_pays": {
+				Type:     schema.TypeBool,
+				Optional: true,
+				Default:  false,
+			},
+			"role_arn": {
+				Type:         schema.TypeString,
+				Required:     true,
+				ForceNew:     true,
+				ValidateFunc: validateArn,
+			},
+			"squash": {
+				Type:     schema.TypeString,
+				Optional: true,
+				Default:  "RootSquash",
+				ValidateFunc: validation.StringInSlice([]string{
+					"AllSquash",
+					"NoSquash",
+					"RootSquash",
+				}, false),
+			},
+		},
+	}
+}
+
+func resourceAwsStorageGatewayNfsFileShareCreate(d *schema.ResourceData, meta interface{}) error {
+	conn := meta.(*AWSClient).storagegatewayconn
+
+	input := &storagegateway.CreateNFSFileShareInput{
+		ClientList:           expandStringSet(d.Get("client_list").(*schema.Set)),
+		ClientToken:          aws.String(resource.UniqueId()),
+		DefaultStorageClass:  aws.String(d.Get("default_storage_class").(string)),
+		GatewayARN:           aws.String(d.Get("gateway_arn").(string)),
+		GuessMIMETypeEnabled: aws.Bool(d.Get("guess_mime_type_enabled").(bool)),
+		KMSEncrypted:         aws.Bool(d.Get("kms_encrypted").(bool)),
+		LocationARN:          aws.String(d.Get("location_arn").(string)),
+		NFSFileShareDefaults: expandStorageGatewayNfsFileShareDefaults(d.Get("nfs_file_share_defaults").([]interface{})),
+		ObjectACL:            aws.String(d.Get("object_acl").(string)),
+		ReadOnly:             aws.Bool(d.Get("read_only").(bool)),
+		RequesterPays:        aws.Bool(d.Get("requester_pays").(bool)),
+		Role:                 aws.String(d.Get("role_arn").(string)),
+		Squash:               aws.String(d.Get("squash").(string)),
+	}
+
+	if v, ok := d.GetOk("kms_key_arn"); ok && v.(string) != "" {
```

Comment:
nit: isn't the `&& v.(string) != ""` unnecessary here? The `ok` from `GetOk` should denote if the value is the zero value. If the config omits `kms_key_arn` then `ok` here will be false, unless I'm not recalling correctly

Reply:
I believe the intention of these `v.(string) != ""` checks with `d.GetOk()` is for allowing modules to define `""` until `null` is available in Terraform 0.12. If this is unnecessary with Terraform 0.11, we have a lot of code to clean up. 🙁